### DISCUSSION TITLE
What does "EXIT: Optimal Solution Found." mean if the last iteration number ends with "r"?

### DISCUSSION THREAD

**Discussion by fahim831** [coin-or/Ipopt#782] [Q&A] [answered]
I have a nonlinear DAE system I've built and discretized with Pyomo and sent to IPOPT to solve. The situation is somewhat similar to [this](https://github.com/coin-or/Ipopt/discussions/568) question.

I suspect it is a badly scaled problem that ends, based on how I do my scaling, with a solved status (i.e., "EXIT: Optimal Solution Found.") or, sometimes, based on other scaling factors, although with the same results for the actual variables, with an infeasible status (the usual expected ending if within the restoration phase). So I am not clear if the problem is really solved (I suspect it is not because some of the variables are off). How do I force IPOPT to exit the restoration phase and what does the current solution mean? Is there some other way to check if it's truly solved?

The problem only has constraints, no objective function, as there are only equations from discretization, which are modeled as constraints in Pyomo.

```
This is Ipopt version 3.13.2, running with linear solver ma27.

Number of nonzeros in equality constraint Jacobian...:    37340
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:     8036

Total number of variables............................:     8678
                     variables with only lower bounds:     3968
                variables with lower and upper bounds:        0
                     variables with only upper bounds:        0
Total number of equality constraints.................:     8678
Total number of inequality constraints...............:        0
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        0

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  1.0000000e+00 1.03e+05 1.00e+00  -1.0 0.00e+00    -  0.00e+00 0.00e+00   0 s
   1  1.0000000e+00 1.03e+03 2.46e+01  -1.0 1.38e+01    -  9.92e-01 9.90e-01h  1 
   2  1.0000000e+00 1.03e+03 2.28e+03  -1.0 8.80e+03    -  2.52e-03 2.48e-03h  1 Nhj 
   3  1.0000000e+00 1.03e+03 3.74e+03  -1.0 1.00e+04    -  4.71e-03 2.47e-05h  1 
   4r 1.0000000e+00 1.03e+03 1.00e+03   3.0 0.00e+00    -  0.00e+00 1.24e-07R  2 
   5r 1.0000000e+00 1.59e+02 1.06e+04   3.0 1.03e+06    -  1.99e-01 9.91e-04f  1 
   6r 1.0000000e+00 2.25e+00 2.84e+03   2.3 7.82e+02    -  9.93e-01 2.02e-01f  1 Nhj
...
76r 1.0000000e+00 4.36e-11 5.38e-07  -9.0 3.21e-05    -  1.00e+00 1.00e+00h  1 

Number of Iterations....: 76

                                   (scaled)                 (unscaled)
Objective...............:   1.0000000000000000e+00    1.0000000000000000e+00
Dual infeasibility......:   0.0000000000000000e+00    0.0000000000000000e+00
Constraint violation....:   4.3602671805767001e-11    4.3602671805767001e-11
Complementarity.........:   0.0000000000000000e+00    0.0000000000000000e+00
Overall NLP error.......:   4.3602671805767001e-11    4.3602671805767001e-11


Number of objective function evaluations             = 79
Number of objective gradient evaluations             = 6
Number of equality constraint evaluations            = 79
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 78
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 76
Total CPU secs in IPOPT (w/o function evaluations)   =      2.658
Total CPU secs in NLP function evaluations           =      0.075

EXIT: Optimal Solution Found.
```

**Comment by svigerske** [comment, marked as answer]
Ipopt went into its feasibility restoration phase and then solved the corresponding NLP (minimize constraint violations). This lead to finding a feasible point (for the non-restoration-phase NLP). Ipopt then checks whether optimality conditions are also satisfied, that is, dual infeasibility and complementarity are below tolerance. Without an objective function, these are of course just zero. And feasible point is optimal here.

>  Is there some other way to check if it's truly solved?

Checking first- and second-order optimality conditions. As there is no objective function, this just means checking that constraints are indeed satisfied. According to Ipopt, constraint violation is at 4e-11, which should be sufficiently good.
